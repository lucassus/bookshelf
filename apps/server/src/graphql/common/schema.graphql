scalar ExternalID

scalar ISODateString

interface Resource {
  id: ExternalID!
}

union Anything = PublicUser | ProtectedUser | Author | Book | BookCopy

interface Timestampable {
  createdAt: ISODateString!
  updatedAt: ISODateString!
}

type Image {
  path: String!
  url: String!
}

interface Error {
  message: String!
}

type MutationError implements Error {
  message: String!
}

type ResourceNotFoundError implements Error {
  message: String!
}

type ValidationError implements Error {
  path: String!
  message: String!
}

type ValidationErrors {
  errors: [ValidationError!]!
}

# TODO: This interface is deprecated, use unions for more typesafe errors handling
interface MutationResponse {
  "A boolean that indicates whether the mutation was successful."
  success: Boolean!

  "Human-readable string that describes the result of the mutation"
  message: String!
}

extend type Query {
  resource(id: ID!): Resource!
  anything(id: ID!): Anything! @deprecated(reason: "No longer supported. Use 'resources' instead")
}
