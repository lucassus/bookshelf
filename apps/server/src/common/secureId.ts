import { SecureId } from "@bookshelf/secure-id";

const secureId = new SecureId<string>({
  separator: "-"
});

export const toInternalIdAndType = (id: string) =>
  secureId.toInternalAndType(id);

export const toInternalId = (id: string): number =>
  Number(secureId.toInternal(id));

export const toExternalId = (entity: { id: number }): string =>
  secureId.toExternal(
    entity.id,
    Object.getPrototypeOf(entity).constructor.name
  );
