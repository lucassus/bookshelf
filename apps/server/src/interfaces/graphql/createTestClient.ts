import { ApolloServer } from "apollo-server-express";
import {
  ApolloServerTestClient,
  createTestClient as createApolloTestClient
} from "apollo-server-testing";
import cookie from "cookie";
import express from "express";
import httpMocks from "node-mocks-http";

import { AUTH_COOKIE_NAME } from "../../infra/config";
import { User } from "../../infra/database/entity";
import { generateAuthToken } from "../../infra/support/authentication";
import { createContext } from "./context";
import { rootSchema } from "./rootSchema";

export function createTestClient({
  currentUser,
  res = httpMocks.createResponse()
}: {
  currentUser?: User;
  res?: express.Response;
} = {}): ApolloServerTestClient {
  const config = {
    schema: rootSchema,
    context: () => {
      const req = httpMocks.createRequest();

      if (currentUser) {
        req.headers = {
          cookie: cookie.serialize(
            AUTH_COOKIE_NAME,
            generateAuthToken(currentUser)
          )
        };
      }

      return createContext({ req, res });
    }
  };

  return createApolloTestClient(new ApolloServer(config));
}
