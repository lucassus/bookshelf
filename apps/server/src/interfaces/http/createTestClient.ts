import cookie from "cookie";
import express from "express";
import request from "supertest";

import { api as apiRoutes } from "./api/routes";
import { AUTH_COOKIE_NAME } from "~/infra/config";
import { User } from "~/infra/database/entity";
import { generateAuthToken } from "~/infra/support/authentication";

export function createTestClient({
  authToken,
  currentUser
}: { currentUser?: User; authToken?: string } = {}) {
  const app = express();

  app.use((req, res, next) => {
    const token = authToken || (currentUser && generateAuthToken(currentUser));

    if (token) {
      req.headers = {
        cookie: cookie.serialize(AUTH_COOKIE_NAME, token)
      };
    }

    next();
  });

  app.use("/api", apiRoutes);

  return request(app);
}
