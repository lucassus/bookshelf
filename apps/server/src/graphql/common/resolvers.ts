import { GraphQLScalarType } from "graphql";

import { secureId } from "../../common/secureId";
import { findAnythingOrFail } from "../../database/findAnythingOrFail";
import { Context } from "../context";
import { Resolvers } from "../resolvers-types.generated";

const resolvers: Resolvers<Context> = {
  Query: {
    resource: (rootValue, { id }, { connection }) =>
      findAnythingOrFail(id, connection),

    anything: (rootValue, { id }, { connection }) =>
      findAnythingOrFail(id, connection)
  },

  ExternalID: new GraphQLScalarType({
    name: "ExternalID",
    parseValue: (value) => secureId.toInternal(value)
  }),

  Timestampable: {
    __resolveType: (timestampable) =>
      Object.getPrototypeOf(timestampable).constructor.name,

    createdAt: (timestampable) => timestampable.createdAt.toISOString(),
    updatedAt: (timestampable) => timestampable.updatedAt.toISOString()
  },

  Resource: {
    __resolveType: (resource) =>
      Object.getPrototypeOf(resource).constructor.name,

    id: (resource) =>
      secureId.toExternal(resource.id, resource.constructor.name)
  },

  Anything: {
    __resolveType: (anything) =>
      Object.getPrototypeOf(anything).constructor.name
  }
};

export default resolvers;
