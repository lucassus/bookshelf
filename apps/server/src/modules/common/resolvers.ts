import { GraphQLScalarType } from "graphql";

import { Resolvers } from "../resolvers-types.generated";

const resolvers: Resolvers = {
  ISODateString: new GraphQLScalarType({
    name: "ISODateString",
    serialize: (value) => (value instanceof Date ? value.toISOString() : null)
  })
};

export default resolvers;
