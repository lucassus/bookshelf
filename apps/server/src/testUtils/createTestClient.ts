import { ApolloServer } from "apollo-server-express";
import {
  ApolloServerTestClient,
  createTestClient as createApolloTestClient
} from "apollo-server-testing";
import express from "express";
import httpMocks from "node-mocks-http";

import { generateAuthToken } from "../common/authentication";
import { AUTH_COOKIE_NAME } from "../config";
import { User } from "../database/entity";
import { buildContext } from "../graphql/context";
import { rootSchema } from "../graphql/rootSchema";

export function createTestClient({
  currentUser,
  res = httpMocks.createResponse()
}: {
  currentUser?: User;
  res?: express.Response;
} = {}): ApolloServerTestClient {
  const config = {
    schema: rootSchema,
    context: () => {
      const req = httpMocks.createRequest();

      if (currentUser) {
        req.cookies = {
          [AUTH_COOKIE_NAME]: generateAuthToken(currentUser)
        };
      }

      return buildContext({ req, res });
    }
  };

  return createApolloTestClient(new ApolloServer(config));
}
