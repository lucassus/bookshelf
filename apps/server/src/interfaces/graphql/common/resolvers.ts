import { GraphQLScalarType } from "graphql";

import { Resolvers } from "~/types/resolvers.generated";

const resolvers: Resolvers = {
  ISODateString: new GraphQLScalarType({
    name: "ISODateString",
    serialize: (value) => (value instanceof Date ? value.toISOString() : null)
  })
};

export default resolvers;
