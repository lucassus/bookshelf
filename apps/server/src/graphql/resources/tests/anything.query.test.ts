import { gql } from "apollo-server-express";

import { toExternalId } from "../../../common/secureId";
import { createTestClient } from "../../../testUtils/createTestClient";
import {
  createAuthor,
  createBook,
  createBookCopy,
  createUser
} from "../../../testUtils/factories";

describe("anything query", () => {
  const GetAnythingQuery = gql`
    query($id: ID!) {
      anything(id: $id) {
        __typename

        ... on Author {
          id
          name
          books {
            ...BookFragment
          }
        }

        ...BookFragment

        ...UserFragment

        ... on BookCopy {
          id
          owner {
            ...UserFragment
          }
          borrower {
            ...UserFragment
          }
          book {
            ...BookFragment
          }
        }
      }
    }

    fragment BookFragment on Book {
      id
      title
      description
      favourite
    }

    fragment UserFragment on User {
      id
      name

      ... on ProtectedUser {
        email
        isAdmin
      }
    }
  `;

  it("fetches Author", async () => {
    // Given
    const author = await createAuthor();
    await createBook({ author });
    await createBook({ author });
    await createBook({ author });

    // When
    const res = await createTestClient().query({
      query: GetAnythingQuery,
      variables: { id: toExternalId(author) }
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).toMatchSnapshot();
  });

  it("fetches Book", async () => {
    // Given
    const book = await createBook();

    // When
    const res = await createTestClient().query({
      query: GetAnythingQuery,
      variables: { id: toExternalId(book) }
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).toMatchSnapshot();
  });

  it("fetches User", async () => {
    // Given
    const user = await createUser();

    // When
    const res = await createTestClient({ currentUser: user }).query({
      query: GetAnythingQuery,
      variables: { id: toExternalId(user) }
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).toMatchObject({
      anything: {
        __typename: "ProtectedUser",
        id: expect.any(String),
        name: "Randolph Satterfield",
        email: "Lauryn.Crist@yahoo.com",
        isAdmin: false
      }
    });
  });

  it("fetches BookCopy", async () => {
    // Given
    const borrower = await createUser();
    const bookCopy = await createBookCopy({ borrower });

    // When
    const res = await createTestClient().query({
      query: GetAnythingQuery,
      variables: { id: toExternalId(bookCopy) }
    });

    // Then
    expect(res.errors).toBe(undefined);
    expect(res.data).toMatchSnapshot();
  });
});

it("fetches with aliases", async () => {
  // Given
  const book = await createBook();
  const author = await createAuthor();
  const user = await createUser();

  // When
  const res = await createTestClient().query({
    query: gql`
      query(
        $bookId: ExternalID!
        $authorId: ExternalID!
        $userId: ExternalID!
      ) {
        something: book(id: $bookId) {
          ... on Book {
            id
            externalId: id
            headline: title
            description
          }
        }

        author(id: $authorId) {
          ... on Author {
            id
            headline: name
            description: bio
          }
        }

        user(id: $userId) {
          ... on User {
            id
            headline: name
            description: info
          }
        }
      }
    `,
    variables: {
      bookId: toExternalId(book),
      authorId: toExternalId(author),
      userId: toExternalId(user)
    }
  });

  // Then
  expect(res.errors).toBe(undefined);
  expect(res.data).toMatchSnapshot();
});
