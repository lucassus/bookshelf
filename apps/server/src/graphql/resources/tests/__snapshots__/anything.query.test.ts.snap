// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`anything query fetches Author 1`] = `
Object {
  "anything": Object {
    "__typename": "Author",
    "books": Array [
      Object {
        "description": "Laborum cupiditate consequatur et voluptatem.",
        "favourite": true,
        "id": "Mi1Cb29r",
        "title": "Et",
      },
      Object {
        "description": "Aut consequatur incidunt blanditiis sunt.",
        "favourite": true,
        "id": "My1Cb29r",
        "title": "In Eos Quae",
      },
      Object {
        "description": "Voluptatem inventore ut quisquam fugiat itaque non alias.",
        "favourite": true,
        "id": "MS1Cb29r",
        "title": "Suscipit Animi Qui Impedit",
      },
    ],
    "id": "MS1BdXRob3I=",
    "name": "Lorene Deckow",
  },
}
`;

exports[`anything query fetches Book 1`] = `
Object {
  "anything": Object {
    "__typename": "Book",
    "description": "Voluptatem inventore ut quisquam fugiat itaque non alias.",
    "favourite": true,
    "id": "MS1Cb29r",
    "title": "Suscipit Animi Qui Impedit",
  },
}
`;

exports[`fetches with aliases 1`] = `
Object {
  "author": Object {
    "description": "Et voluptatem nostrum.",
    "headline": "Suzanne Moore IV",
    "id": "Mi1BdXRob3I=",
  },
  "something": Object {
    "description": "Voluptatem inventore ut quisquam fugiat itaque non alias.",
    "externalId": "MS1Cb29r",
    "headline": "Suscipit Animi Qui Impedit",
    "id": "MS1Cb29r",
  },
  "user": Object {
    "description": "Sint voluptatibus deserunt totam quae eos.",
    "headline": "Micheal Wiza III",
    "id": "MS1Vc2Vy",
  },
}
`;
